File: android-beans/src/main/java/rocks/frieler/android/beans/scopes/prototype/PrototypeScopedFactoryBeanHandler.java
package rocks.frieler.android.beans.scopes.prototype;

import rocks.frieler.android.beans.scopes.ScopedFactoryBean;
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler;

/**
 * {@link ScopedFactoryBeanHandler} for the {@value #PROTOTYPE_SCOPE}-scope.
 * <p>
 * It produces a new bean-instance every time.
 */
public class PrototypeScopedFactoryBeanHandler implements ScopedFactoryBeanHandler {
    public static final String PROTOTYPE_SCOPE = "prototype";

    @Override
    public String getName() {
        return PROTOTYPE_SCOPE;
    }

    @Override
    public boolean isActive() {
        return true;
    }

    @Override
    public <T> T getBean(String name, ScopedFactoryBean<T> factoryBean) {
        return factoryBean.produceBean();
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/scopes/prototype/PrototypeScopedFactoryBeanHandler.kt
package rocks.frieler.android.beans.scopes.prototype

import rocks.frieler.android.beans.BeansProvider
import rocks.frieler.android.beans.scopes.ScopedFactoryBean
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler

/**
 * [ScopedFactoryBeanHandler] for the {@value #PROTOTYPE_SCOPE}-scope.
 *
 *
 * It produces a new bean-instance every time.
 */
class PrototypeScopedFactoryBeanHandler : ScopedFactoryBeanHandler {
    override val name: String
        get() = PROTOTYPE_SCOPE

    override val isActive: Boolean
        get() = true

    override fun <T :Any> getBean(name: String, factoryBean: ScopedFactoryBean<T>, dependencies: BeansProvider): T {
        return factoryBean.produceBean(dependencies)
    }

    companion object {
        const val PROTOTYPE_SCOPE = "prototype"
    }
}
